src/histbuf.rs

```diff
@@ -0,0 +1,325 @@
+    /// ```
+    pub fn last(&self) -> &T {
+        &self.data[(self.write_at + self.len() - 1) % self.len()]
+    }
+
+    /// Returns an iterator over the elements of the buffer.
+    ///
+    /// Note: if the order of elements is not important, use
+    /// `.as_slice().iter()` instead.
+    pub fn iter(&self) -> Iter<'_, T> {
+        Iter {
+            data: &self.data,
+            cur: self.write_at,
+            left: self.len(),
+        }
+    }
+
+    /// Returns an iterator over mutable elements of the buffer.
+    ///
+    /// Note: if the order of elements is not important, use
+    /// `.as_mut_slice().iter_mut()` instead.
+    pub fn iter_mut(&mut self) -> impl Iterator<Item = &'_ mut T> {
+        let (p1, p2) = self.data.split_at_mut(self.write_at);
+        p2.iter_mut().chain(p1)
+    }
+}
+
+pub struct Iter<'a, T> {
+    data: &'a [T],
+    cur: usize,
+    left: usize,
+}
+
+impl<'a, T> Iterator for Iter<'a, T> {
+    type Item = &'a T;
+
+    fn next(&mut self) -> Option<&'a T> {
+        if self.left == 0 {
+            None
+        } else {
+            let el = &self.data[self.cur];
+            self.cur = (self.cur + 1) % self.data.len();
+            self.left -= 1;
+            Some(el)
+        }
+    }
+}
+
+impl<T, N> Extend<T> for HistoryBuffer<T, N>
+where
+    N: ArrayLength<T>,
+{
+    fn extend<I>(&mut self, iter: I)
+    where
+        I: IntoIterator<Item = T>,
+    {
+        for item in iter.into_iter() {
+            self.push(item);
+        }
+    }
+}
+
+impl<'a, T, N> Extend<&'a T> for HistoryBuffer<T, N>
+where
+    T: 'a + Clone,
+    N: ArrayLength<T>,
+{
+    fn extend<I>(&mut self, iter: I)
+    where
+        I: IntoIterator<Item = &'a T>,
+    {
+        self.extend(iter.into_iter().cloned())
+    }
+}
+
+impl<T, N> AsSlice for HistoryBuffer<T, N>
+where
+    N: ArrayLength<T>,
+{
+    type Element = T;
+
+    /// Returns the array slice backing the buffer, without keeping track
+    /// of the write position. Therefore, the element order is unspecified.
+    fn as_slice(&self) -> &[T] {
+        &self.data
+    }
+}
+
+impl<T, N> AsMutSlice for HistoryBuffer<T, N>
+where
+    N: ArrayLength<T>,
+{
+    /// Returns the array slice backing the buffer, without keeping track
+    /// of the write position. Therefore, the element order is unspecified.
+    fn as_mut_slice(&mut self) -> &mut [T] {
+        &mut self.data
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::{consts::*, HistoryBuffer};
+    use as_slice::{AsSlice, AsMutSlice};
+
+    #[test]
+    fn new() {
+        let x: HistoryBuffer<u8, U4> = HistoryBuffer::new_with(1);
+        assert!(x.iter().eq([1; 4].iter()));
+
+        let x: HistoryBuffer<u8, U4> = HistoryBuffer::new();
+        assert!(x.iter().eq([0; 4].iter()));
+    }
+
+    #[test]
+    fn push_iter() {
+        let mut x: HistoryBuffer<u8, U4> = HistoryBuffer::new();
+        x.push(1);
+        x.push(4);
+        assert!(x.iter().eq([0, 0, 1, 4].iter()));
```

Comment:
Related to my previous question about `len` returning not always the capacity, I guess. Would someone expect this to iterate from the oldest pushed element to the newest pushed element, *not* including the not-pushed elements (i.e. the elements that have the `Default` value)? e.g. this iteration should yield `1` and then `4` and then `None`. Or to word this differently: I didn't expect this iteration to start with the unpushed / cleared values.